external-plugins/rehearse/plugin/main.go

```diff
@@ -0,0 +1,216 @@
+	fs.StringVar(&o.jobsNs,
+		"jobs-namespace",
+		"",
+		"The namespace in which Prow jobs should be created.")
+	fs.BoolVar(&o.alwaysRun,
+		"always-run",
+		false,
+		"If set to true, will act on new and updated PRs. Otherwise, act only on comments.")
+	fs.StringVar(&o.cacheDir,
+		"cache-dir",
+		"",
+		"Directory to store git repos cache in.")
+	for _, group := range []flagutil.OptionGroup{&o.github} {
+		group.AddFlags(fs)
+	}
+	fs.Parse(os.Args[1:])
+	return o
+}
+
+func clientFactoryCacheDirOpt(cacheDir string) func(opts *git.ClientFactoryOpts) {
+	return func(cfo *git.ClientFactoryOpts) {
+		cfo.CacheDirBase = &cacheDir
+	}
+}
+
+func main() {
+	opts := gatherOptions()
+	opts.validate()
+
+	logger := setupLogger()
+	logger.Infoln("Setting up events server")
+
+	var config *rest.Config
+	var err error
+	if opts.kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", opts.kubeconfig)
+		mustSucceed(err, "Could not instantiate K8s config from the given kubeconfig.")
+	} else {
+		config, err = rest.InClusterConfig()
+		mustSucceed(err, "Could not instantiate K8s config from the in cluster config.")
+	}
+	prowClient, err := v1.NewForConfig(config)
+	mustSucceed(err, "Could not instantiate a Prow client from the given kubeconfig.")
+
+	secretsAgent := startSecretsAgent(opts.github.TokenPath, opts.hmacSecretFile)
+
+	githubClient, err := opts.github.GitHubClient(secretsAgent, opts.dryRun)
+	mustSucceed(err, "Could not instantiate github client.")
+
+	gitClientFactory, err := git.NewClientFactory(clientFactoryCacheDirOpt(opts.cacheDir))
+	mustSucceed(err, "Could not instantiate git client factory")
+
+	eventsChan := make(chan *handler.GitHubEvent)
+
+	eventsServer := server.NewGitHubEventsServer(
+		secretsAgent.GetTokenGenerator(opts.hmacSecretFile),
+		eventsChan)
+	eventsHandler := handler.NewGitHubEventsHandler(
+		eventsChan,
+		logger,
+		prowClient.ProwJobs(opts.jobsNs),
+		githubClient,
+		opts.prowConfigPath,
+		opts.jobsConfigBase,
+		opts.alwaysRun,
+		gitClientFactory)
+
+	// run the events server
+	serverMux := http.NewServeMux()
+	serverMux.Handle(opts.endpoint, eventsServer)
+	srv := &http.Server{Addr: fmt.Sprintf(":%d", opts.port), Handler: serverMux}
+	stopChan := make(chan bool, 1)
+	go eventsHandler.Start(stopChan)
+	interrupts.ListenAndServe(srv, 5*time.Second)
+	logger.Infoln("Events server is listening on port:", opts.port)
+	externalplugins.ServeExternalPluginHelp(serverMux, logger.WithField("plugin-help", ""), helpProvider)
+	interrupts.WaitForGracefulShutdown()
+	stopChan <- true
+	logger.Println("Rehearse server was gracefully shut down")
+}
+
+func helpProvider(_ []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: `The rehearse plugin is used run modified Prow jobs from a PR to provide pre-merge feedback.`,
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Usage:       "Send a PR or type /rehearse on the PRs page",
```

Comment:
Right, I've added just the permitted label for now because I think that it keeps things simpler.
You either permit or you don't. And if we will want rehearsals for jobs, we will add them as we modify them... I think it should work just fine.